kratos/factories/base_factory.h

```diff
@@ -0,0 +1,283 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Vicente Mataix Ferrandiz
+//
+
+#if !defined(KRATOS_BASE_FACTORY_H_INCLUDED )
+#define  KRATOS_BASE_FACTORY_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/kratos_parameters.h"
+#include "includes/kratos_components.h"
+#include "includes/shared_pointers.h"
+#include "linear_solvers/linear_solver.h"
+#include "solving_strategies/schemes/scheme.h"
+#include "solving_strategies/strategies/solving_strategy.h"
+#include "solving_strategies/builder_and_solvers/builder_and_solver.h"
+#include "solving_strategies/convergencecriterias/convergence_criteria.h"
+#include "spaces/ublas_space.h"
+
+namespace Kratos
+{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class BaseFactory
+ * @ingroup KratosCore
+ * @brief Here we add the functions needed for the registration of class
+ * @details Defines the base class factory
+ * @author Vicente Mataix Ferrandiz
+ * @tparam TClass The class to create the factory
+ * @tparam TAuxiliarClass The auxiliar class to create the factory
+ */
+template<typename TClass, typename TAuxiliarClass = TClass>
```

Comment:
See @pooyan-dadvand this is closer to the second option you mentioned

Reply:
I said **variadic** template argument.

Reply:
Ok I din't know that that existed, cool

Reply:
I am studying this and it looks like a little bit of dark magic

Reply:
Somehow it is... but here would be useful

Reply:
https://en.cppreference.com/w/cpp/language/parameter_pack

I understand the concept, but I am not sure how to use it when I need to know the types for certain functions... Maybe creating private functions, and deduce automatically, but maybe not possible

Reply:
I cannot make it work, with the multiple hesitance I have problems to handle it. Do we really need to use the variadic template?, right now with two template arguments is necessary for most of the classes we use